cmd/kubermatic-installer/cmd_local.go

```diff
@@ -0,0 +1,400 @@
+
+	ms := kubermaticmaster.MasterStack{}
+	k, uk, err := loadKubermaticConfiguration(kubermaticPath)
+	if err != nil {
+		logger.Panicf("Failed to load %v after autoconfiguration: %v", kubermaticPath, err)
+	}
+	v, err := loadHelmValues(valuesPath)
+	if err != nil {
+		logger.Panicf("Failed to load %v after autoconfiguration: %v", valuesPath, err)
+	}
+	msOpts := stack.DeployOptions{
+		ChartsDirectory:            opts.ChartsDirectory,
+		KubeClient:                 kubeClient,
+		HelmClient:                 helmClient,
+		Logger:                     log.Prefix(logrus.NewEntry(logger), "   "),
+		MLASkipMinio:               true,
+		MLASkipMinioLifecycleMgr:   true,
+		KubermaticConfiguration:    k,
+		RawKubermaticConfiguration: uk,
+		HelmValues:                 v,
+	}
+
+	if err := ms.Deploy(context.Background(), msOpts); err != nil {
+		logger.Fatalf("Failed to deploy kubermatic stack: %v", err)
+	}
+
+	kubeconfig := filepath.Join(dir, "kube-config.yaml")
+	internalKubeconfig := filepath.Join(dir, "kube-config-internal.yaml")
+	kindSeedSecret := initKindSeedSecret(kubeClient, logger, kubeconfig, internalKubeconfig)
+	ensureResource(kubeClient, logger, &kindSeedSecret)
+	kindPreset := initKindPreset(logger, internalKubeconfig)
+	ensureResource(kubeClient, logger, &kindPreset)
+	ensureResource(kubeClient, logger, &kindLocalSeed)
+	return kkpEndpoint
+}
+
+func localKindFunc(logger *logrus.Logger, opt LocalOptions) cobraFuncE {
+	return handleErrors(logger, func(cmd *cobra.Command, args []string) error {
+		dir := "./examples"
+		path, err := os.Executable()
+		if err == nil {
+			dir = filepath.Join(filepath.Dir(path), "examples")
+		}
+		kubeClient, cancel := localKind(logger, dir)
+		defer cancel()
+		kubeconfig := filepath.Join(dir, "kube-config.yaml")
+		helmClient, err := helm.NewCLI(opt.HelmBinary, kubeconfig, "", opt.HelmTimeout, logger)
+		if err != nil {
+			logger.Fatalf("Failed to create helm client: %v", err)
+		}
+		helmVersion, err := helmClient.Version()
+		if err != nil {
+			logger.Fatalf("Failed to check Helm version: %v", err)
+		}
+		if helmVersion.LessThan(MinHelmVersion) {
+			logger.Fatalf(
+				"the installer requires Helm >= %s, but detected %q as %s (use --helm-binary or $HELM_BINARY to override)",
+				MinHelmVersion,
+				opt.HelmBinary,
+				helmVersion,
+			)
+		}
+		installKubevirt(logger, dir, helmClient, opt)
+		endpoint := installKubermatic(logger, dir, kubeClient, helmClient, opt)
+		logger.Infof("KKP installed successfully, login at http://%v", endpoint)
+		return nil
+	})
+}
+
+func getLocalIP(logger *logrus.Logger) string {
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		logger.Fatalf("unable to get interfaces: %v", err)
+	}
+	for _, address := range addrs {
+		if ipnet, ok := address.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
+			if ipnet.IP.To4() != nil {
+				return ipnet.IP.String()
+			}
+		}
+	}
+	logger.Fatalf("no interface with public IP")
+	return ""
+}
```

Comment:
This function is very wrong. You need to ether take IP from the CLI flag and fallback to the IP of the interface which is used as a default gateway.

Reply:
```go
package main

import (
	"fmt"
	"net"
)

func main() {
	fmt.Printf("local ip: %q\n", getLocalIP())
}

func getLocalIP() string {
	addrs, err := net.InterfaceAddrs()
	if err != nil {
		panic(fmt.Errorf("unable to get interfaces: %v", err))
	}

	for _, address := range addrs {
		if ipnet, ok := address.(*net.IPNet); ok && !ipnet.IP.IsLoopback() {
			if ipnet.IP.To4() != nil {
				return ipnet.IP.String()
			}
		}
	}

	panic(fmt.Errorf("no interface with public IP"))

	return ""
}
```

this will return me
```
local ip: "172.18.0.1"
```
but `ip route | grep default` will show
```bask
$ ip route | grep default
default via 192.168.18.1 dev enp0s20f0u6u3c2 proto dhcp src 192.168.18.40 metric 100
```

Reply:
that is a good point, any opinion on using https://github.com/jackpal/gateway for this? I haven't found any OS-independent way in the go core packages for this. Also, I would like to keep as fallback the algorithm to find a first interface address that is ipv4 and not loopback. And of course, allow users to bypass the autoconfiguration by passing `--endpoint=172.18.0.1.nip.io` for example.

Reply:
Using `gateway.DiscoverInterface()` returns the correct result. But I don't see the point of using the fallback in the form of "just take the first one of them all". It should be ether `gateway.DiscoverInterface()` (and then empty `--endpoint=` is OK) or `--endpoint=` should be provided explicitly.

Reply:
re-implemented in https://github.com/kubermatic/kubermatic/pull/12216/commits/bb80ea4a228766a645a88362872cfc4dfad16c4, let me know if you have additional suggestions